### DISCUSSION TITLE
pre-push 훅에 기대하는 동작이 궁금합니다

### DISCUSSION THREAD

**Discussion by jiyeon2** [wanted-pre-onboarding-side-project/habit-tracker#36] [Q&A] [answered]
#35 pr메시지 하단 하단에서 언급한 내용입니다

prettier가 포매팅을 처리해주는 pre-commit훅과 달리
pre-push 훅의 경우
단순히 `eslint --cache .`  을 실행하고 있어서
eslint가 자동으로 수정가능한 오류를 고치지 않으며
lint 에러가 있는경우 오류만 출력하고 push가 진행되지 않습니다

즉, pre-push 훅은 lint에러가 있는 코드는 push되지 않도록 막는 역할을 하고 있습니다
개발자가 직접 lint에러를 수정하고, stage에 올리고, commit을 진행해야 합니다

pre-commit에서 prettier가 자동 포맷팅한 코드가 commit에 반영되면 좋겠다는 생각에
pre-push에 실행할 명령어를 `eslint --cache --fix .` 로 변경해서 실행해 본 결과
eslint가 자동수정은 해주지만, 수정된 파일은 기존 commit에 반영되어 있지 않으므로
기존에 commit된, lint오류가 존재하는 코드가 push되는 것을 확인하였습니다

pre-push 훅에서 lint에러 자동수정된 코드를 staging, commit 하려면
상연님이 작성하신 pre-commit 파일 코드처럼 추가 명령어 작성이 필요할 것으로 보이는데
제가 그에 대한 지식이 부족하여 시도하지 못했습니다

저는 vscode에서 저장시 eslint 자동수정 기능을 켜놓고 있어서 
push 단계에서는 eslint가 수정하지 못한 린트오류만 남습니다
그래서 pre-push 훅에서 제가 확인하지 못한 린트오류가 있다는 것을 알려주고 
push를 막아주는 것만으로도 충분합니다

여러분들도 pre-push 훅에서 린트오류가 있는 경우 에러만 출력하는 것으로 괜찮으신지 궁금합니다~
그리고 혹시 pre-push 훅을 개선하는 방법을 알고계시거나 
관련 지식이 있으시면 공유해주셨으면 합니다

**Comment by chadseok** [comment]
네 인턴십 기간동안 그런 문제에 대해서 저희 팀원분중에 얘기를 하신분이 있었습니다. 저희 팀은 그 때 `lint-staged`라는 라이브러리를 사용하여 해결하였습니다. 그 라이브러리에서 **eslint에서 fix된 코드까지 다시 git 내용에 포함시켜주는 기능**이 있어서 사용하였습니다. 

그래서 이 부분에 대해서 저도 여쭤볼려고 생각은 하고 있었습니다. 다만 `.husky/pre-commit`에 적힌 스크립트에서 그런 기능을 자체적으로 해주는 것 같아서 `.husky/pre-push`에도 비슷하게 작성하면 되지 않을 까 생각합니다. 저는 일단 스크립트 작성과 관련된 지식이 없기도 해서 라이브러리를 사용하였습니다

**Comment by greyHairChooseLife** [comment, marked as answer]
저는 eslint의 `fix` 옵션을 사용 하지 않는 것이 어떨까 합니다. 

물론 지정한 규칙들에 맞게, 린터가 적어도 개인보다는 훨씬 적절하게 고쳐 줄 것입니다. 그러나 코딩 컨벤션(규칙)으로 나와 있다는 것은 언어나 라이브러리에서 강제할 정도는 아니어도 개발자가 **알고** 있으면 좋은 것이라고 생각합니다. 

`fix` 옵션 없이 사용하면 자신의 코드에 어떤 warning과 error가 있는지 모두 확인 할 수 있고요, 직접 간단히 수정 해 가면서 적절한 습관을 들일 수 있습니다.

물론 제가 말하는 것은 선택 사항이지만, 그렇기에 이 부분은 공통적으로 적용되는 git hook 보다는 npx로 직접 실행하던가, 지연님처럼 코드 에디터의 자동 적용 기능을 활용하던가, 또는 `package.json`에 스크립트를 정의 해 편하게 사용하는 정도가 적절할 것 같습니다.

더군다나 `pre-push` 깃 훅에 `fix`를 사용한다면 수정 된 내용을 직접 확인하는 과정 없이 곧바로 `push`됩니다. 비록 린터 규칙에 대한 것 정도지만 자신의 코드가 정확히 어떤 상태인지 모르고 `push`하는 것은 개발자의 좋은 습관이 아니라고 생각합니다.

---

참고로 저희 팀은 `pre-push`도 직접 만들어서 썼는데요, `fix`는 쓰지 않고 warning이 남아 있음에도 정말로 `push` 할 것인지 CLI로 인터렉티브하게 확인하는 단계를 넣었습니다. warning이 있는 것 자체가 너무 빡빡하게 규칙을 단속해서 개발이 불편한 것을 막는 용도인데, 그러다보니 최종 결과물 제출 할 때에도 깜빡 잊고 내버려둔 적이 있어서요.

혹시 이런 걸 도입 해 보시겠다면 제가 약간 수정해서 PR 해 보겠습니다.

**Reply by jiyeon2** [reply]
fix 옵션을 안쓰는 편이 저도 좋습니다~
지금 불편함이 없는데 굳이 @greyHairChooseLife 님이 추가 작업을 할 필요는 없어 보입니다!
두분 다 답변 감사합니다!
